### ISSUE TITLE
"unable to load DLL 'Microsoft.Data.SqlClient.SNI.dll'" errors with pwsh and credentials

### ISSUE THREAD

**Issue by paschott** [dataplat/dbatools#8268]
### Verified issue does not already exist?

Yes

### What error did you receive?

Started getting "unable to load DLL 'Microsoft.Data.SqlClient.SNI.dll' " errors w/ dbatools in pwsh when trying to run commands using credentials.

e.g., `Get-DbaMaxMemory machinename -SqlCredential $cred` throws that error.

I seem to get it whenever passing in a credential w/ a dbatools function, but that same credential works for a native command like "Enter-PSSession".

### Steps to Reproduce

```$cred = get-credential domain\user
Get-DbaMaxMemory machinename -SqlCredential $cred
```
Get "unable to load DLL 'Microsoft.Data.SqlClient.SNI.dll' " error

### Are you running the latest release?

Yes

### Other details or mentions

This command worked within the last month or so as I use dbatools functions pretty regularly and have to pass in alternate credentials to connect to various machines.  Here are the changes I can think of:
SSMS v18.11.1 upgrade
dbatools upgrade (to latest releases)
pwsh.exe upgrade to v7.2.2

These functions work on the same machine using Windows PowerShell v5.1

### What PowerShell host was used when producing this error

PowerShell Core (pwsh.exe)

### PowerShell Host Version

Name                           Value
----                           -----
PSVersion                      7.2.2
PSEdition                      Core
GitCommitId                    7.2.2
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

### SQL Server Edition and Build number

n/a

### .NET Framework Version

.NET 6.0.3

**Comment by paschott** [comment]
Quick note - `Get-DbaDiskSpace` seems to work with the -Credential parameter, but `Get-DbaMaxMemory` does not work with a credential parameter.  Tested using the same credential and the same target machine using pwsh.

**Comment by paschott** [comment]
Same behavior w/ v1.1.86 and pwsh.exe.

**Comment by paschott** [comment]
I was able to load the file in the dbatools bin\smo folder by using "add-type" for the file, but still get an Access Denied error when trying to use the `-SqlCredential` option with dbatools.

**Comment by andreasjordan** [comment]
Hi @paschott - can you still reproduce the issue?

**Comment by paschott** [comment]
@andreasjordan I ended up working around it by granting permissions to that "admin" account for the PS Modules folder, so I stopped getting the error there. That's the only machine I really use that for right now so the problem is mitigated, but that's with a workaround.  I'll have to try removing permissions from that folder to see if I'm still getting the error w/ the latest pwsh + dbatools module.

**Comment by andreasjordan** [comment]
Thanks for the feedback. I will leave this open for a while, happy to hear from you.

**Comment by potatoqualitee** [comment]
I think the new library in the upcoming dbatools 2.0 will suit you a lot better. no idea why the admin thing worked (cuz i dont understand dlls) but im doing it very differently and it should be more reliable.

**Comment by paschott** [comment]
Well, it was some way it handled the permissions. If I called a dbatools function that queried some server but used credentials. For some reason, those credentials seemed to be used to access my _local_ files. Once I granted permissions to that credential account to my local folder, the calls worked. I hadn't thought about it too much since making that tweak.  Of course, that wouldn't work for cases where the domain isn't trusted or similar.  I'll re-test shortly once I get over to that machine.

**Comment by potatoqualitee** [comment]
Ahhhh, okay. There will be nothing I can do on my end -- I never considered this as a side effect of installing to the current user scope! Please continue with this workaround or install to the system scope.

**Comment by wsmelton** [comment]
@potatoqualitee @andreasjordan
I can confirm this issue still exists even in the new code changes coming in 2.0 release.

The issue that is happening is once a credential is provided PowerShell is performing an impersonation trying to create the DLL file in a location within the user's PATH environemnt. You can observe in the screenshot below all the CreateFile operations occurring where it is going through the paths in my PATH environment trying to create that DLL file.

![image](https://user-images.githubusercontent.com/11204251/208729724-31831e8a-5b20-47b4-9eb0-cfdd571ea4de.png)

This is not normal but I'm not aware of why it is happening yet.

**Comment by potatoqualitee** [comment]
I opened up PowerShell and did nothing (I was planning to import SqlServer) and had similar results

![image](https://user-images.githubusercontent.com/8278033/208899658-d5eb1026-eebb-435c-a791-18d3a2a8797a.png)

Weirdly, when I import SqlServer, i get no hits for SNI. I just have no clue how they import things but I imagine it's drastically different because everything is compiled.

Ultimately, the permission denied thing makes sense to me if the module is installed in the user scope. `Enter-PSSession` works because it's available in the global scope. Get-DbaDiskSpace uses purely remotely available calls and can be run against non-SQL Servers, while Test-DbaMaxMemory makes a call to SQL Server and thus needs the SqlClient DLLs. If this worked previously, it's because we used System.Data and not Microsoft.Data. System.Data is available system-wide whereas Microsoft.Data is available in our module. Ultimately, I'd say to install with the AllUsers scope and there should be no issue.

**Comment by potatoqualitee** [comment]
Oh also, the CreateFile is a misnomer if I had to guess. I think it is actually just trynna open it.

**Comment by wsmelton** [comment]
All user scope won't fix this because I'm using the module from a saved path. Recall I'm using the development branch of GitHub (local development) so the module exist under a completely separate path that is not even part of the PSModule paths.

**Comment by potatoqualitee** [comment]
Yes, that account is not aware of that path which alluserscope would fix, I'm guessing. testing now!

**Comment by potatoqualitee** [comment]
I added the github path via registry to all users to emulate alluser scope and tested with success

![image](https://user-images.githubusercontent.com/8278033/208930624-02ecfa95-d1bb-4443-b84e-7a0caa49760b.png)

AW BRUH, I did this all with powershell.exe. BRB

**Comment by potatoqualitee** [comment]
no idea how to work around this :( anyone know more about interoperability?

![image](https://user-images.githubusercontent.com/8278033/208932688-0bab4f73-85ed-4f52-bb7f-7d8524b79ba1.png)

**Comment by wsmelton** [comment]
It is only the core library module that has to be installed in all user scope, follow now. That is a pretty big breaking change for those users that don't use `Install-Module` or `Install-PSResource` in their environment.

**Comment by potatoqualitee** [comment]
yeah, i drew the line there and wont support install.ps1 anymore. they'll have to find a machine with install-module or save-module or download the zips manually.

**Comment by potatoqualitee** [comment]
This will need some docs if that is the solution. Or like, figure out how to do 7 to 7 without much fuss. Any idea?

**Comment by niphlod** [comment]
![image](https://user-images.githubusercontent.com/122119/208947826-a924d7d9-e9a8-4f5e-ba2c-47b5601810b6.png)
 ?

**Comment by niphlod** [comment]
nope but .... https://github.com/PowerShell/PowerShell/issues/11616

**Comment by wsmelton** [comment]
Not sure remoting plays into here because there is no session created.

**Comment by potatoqualitee** [comment]
I dont believe you can do `Invoke-Command -ComputerName localhost -Credential $cred` without remoting being enabled. I had to enable it for my machine. Maybe the underlying framework (or we) skips it? or doesnt skip it?

**Comment by potatoqualitee** [comment]
but a session must be created to downgrade to 5.1 from 7? ohh lemme check the pid

**Comment by potatoqualitee** [comment]
innnnnnnnnnnnnterresting

```
[7ms] C:/github❯ $pid
11952
[1ms] C:/github❯ Invoke-Command -ComputerName workstation -Credential $cred -ScriptBlock { $pid }
8684
[272ms] C:/github❯ Invoke-Command -ComputerName workstation -Credential $cred -ScriptBlock { $pid }
20568
```

**Comment by wsmelton** [comment]
If a user wants to run the module in Windows PowerShell from PS 7 on their local machine they can `Import-Module ... -UseWindowsPowerShell`. That will create a runspace in PS7 session with a remote session to the 5.1 host on their local machine.

**Comment by potatoqualitee** [comment]
Unfortunately, that's probably not going to be too helpful.

```
PowerShell 7.3.1
PS C:\github> Import-Module -UseWindowsPowerShell dbatools
WARNING: Module dbatools is loaded in Windows PowerShell using WinPSCompatSession remoting session; please note that all input and output of commands from this module will be deserialized objects. If you want to load this module into PowerShell please use 'Import-Module -SkipEditionCheck' syntax.
PS C:\github> Connect-DbaInstance -SqlInstance sqlcs

ComputerName Name  Product Version   HostPlatform IsAzure IsClustered ConnectedAs
------------ ----  ------- -------   ------------ ------- ----------- -----------
             sqlcs         15.0.4236
Out-Default: urn could not be resolved at level OleDbProviderSetting.
```

**Comment by wsmelton** [comment]
I'm using 7.3.1 and whether I'm passing in a credential or creating a connection with credential I don't see the issue occurring with PS Remoting.

**Comment by potatoqualitee** [comment]
any idea how to fix this, anyone?

**Comment by andreasjordan** [comment]
Is this still a problem in version 2.0?

**Comment by dominic-burns** [comment]
This caused problems for me too. 

I had initially installed dbatools v2.0.4 to PS 7.3.6 with no scope specified: 
install-module dbatools
It installed to  _C:\Users\<user>\Documents\PowerShell\Modules\dbatools_
I had the same problems as described by other submitters - when specifying credentials it would complain about the missing DLL. (The user with the specified credentials did not have permissions to read the folders under c:\users)

I then uninstalled and installed for all users: 
**install-module dbatools -Scope AllUsers**
However, the same missing DLL errors occurred.

Finally I realised that the **C:\Users\\\<user>\Documents\PowerShell\Modules\dbatools** folder had not been automatically removed as part of the original uninstall. This folder is normally be removed at uninstall.

I removed this folder and the issue was resolved. 

My assumption is that dbatools will search in that folder first for the dll and then have the same permission issue as the 'CurrentUser' scope and fail. If the folder does not exist it instead searches in the correct global powershell module folder where it has permission to read the DLL.

Hope this helps!

**Comment by wsmelton** [comment]
Still an issue even in PowerShell 7.4 😢

**Comment by mashassanpour** [comment]
Hi All,

I managed to fix my issue this way:

`PS C:\> Get-Module -ListAvailable | Where-Object -Property Name -Like *dba*

    Directory: C:\powershell\Modules\offline

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     2.1.22                dbatools                            Desk      {Select-DbaObject, Set-DbatoolsConfig, Test-DbaAgSpn, Install-DbaAgentAdminAlert…}
Script     2024.4.12             dbatools.library                    Desk      Get-DbatoolsLibraryPath`

then I removed the module:

`PS C:\> uninstall-Module -name dbatools.library
PS C:\> uninstall-Module -name dbatools`

then I installed the module using offline mode. you can find the instructions [here](https://blog.netnerds.net/2023/04/offline-install-of-dbatools-and-dbatools-library/).

`PS C:\> Save-Module dbatools -Path C:\Powershell\Modules\offline
PS C:\> Import-Module C:\Powershell\Modules\offline\dbatools.library
PS C:\> Import-Module C:\Powershell\Modules\offline\dbatools
PS C:\> (Get-Module -ListAvailable | Where-Object -Property Name -Like *dba*).Path
C:\powershell\Modules\offline\dbatools\2.1.22\dbatools.psd1
C:\powershell\Modules\offline\dbatools.library\2024.4.12\dbatools.library.psd1`

Finally I ran a test:

`PS C:\> $sqladmin = Get-Credential -UserName  masouds1

PS C:\> Get-DbaDatabase -SqlInstance scafell\WI -SqlCredential $sqladmin -ExcludeSystem | Select-Object -Property name | Format-Table

Name
----
vCenter6.0
VUM6.0
vSphere
VUM
Aurius
ToolsDB`

I hope you find it useful.

**Comment by mattcargile** [comment]
`-SqlCredential` works in _PowerShell_ `7.5.2` and `2.5.5` of `dbatools`. `5.1` still throws that error in an interactive session an a _PSSession_.
